### PR TITLE
Make spacegrep export json for semgrep

### PR DESCRIPTION
**Pull request by mjambon**
The command-line option is `--output-format semgrep` (see `spacegrep --help`). There's a good chance that the json format is wrong in some places, since I just guessed it from the ocaml implementation.

Here's some sample output (searching for the pattern `free $X` in the text of the GNU AGPL license):

```
~/semgrep/spacegrep $ ./bin/spacegrep 'free $X' -d LICENSE --output-format semgrep
{
  "matches": [
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 18, "col": 69, "offset": 828 },
      "end": { "line": 19, "col": 9, "offset": 841 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 19, "col": 1, "offset": 833 },
            "end": { "line": 19, "col": 9, "offset": 841 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "1a629dbf487ee1ca79cc2d1e486ce02c"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 21, "col": 20, "offset": 881 },
      "end": { "line": 21, "col": 33, "offset": 894 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 21, "col": 25, "offset": 886 },
            "end": { "line": 21, "col": 33, "offset": 894 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "bbc57b936bbfa438158dc98783674d59"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 23, "col": 42, "offset": 1041 },
      "end": { "line": 23, "col": 55, "offset": 1054 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 23, "col": 47, "offset": 1046 },
            "end": { "line": 23, "col": 55, "offset": 1054 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "5ce1e82ba71603d64920df43efd511ab"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 26, "col": 1, "offset": 1209 },
      "end": { "line": 26, "col": 14, "offset": 1222 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 26, "col": 6, "offset": 1214 },
            "end": { "line": 26, "col": 14, "offset": 1222 },
            "abstract_content": "programs",
            "unique_id": {
              "type": "AST",
              "md5sum": "16a2dfd24543a2ac106890f59cde43b4"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 36, "col": 34, "offset": 1729 },
      "end": { "line": 36, "col": 47, "offset": 1742 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 36, "col": 39, "offset": 1734 },
            "end": { "line": 36, "col": 47, "offset": 1742 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "d0df4dcb9761c250dc5958a221841740"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 127, "col": 67, "offset": 6280 },
      "end": { "line": 128, "col": 9, "offset": 6293 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 128, "col": 1, "offset": 6285 },
            "end": { "line": 128, "col": 9, "offset": 6293 },
            "abstract_content": "programs",
            "unique_id": {
              "type": "AST",
              "md5sum": "20cf165194aaee276b15967c27d7c83a"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 490, "col": 10, "offset": 25228 },
      "end": { "line": 490, "col": 17, "offset": 25235 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 490, "col": 15, "offset": 25233 },
            "end": { "line": 490, "col": 17, "offset": 25235 },
            "abstract_content": "of",
            "unique_id": {
              "type": "AST",
              "md5sum": "bb0abe8681c10329a1fa96f8504ea7b7"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 626, "col": 1, "offset": 32588 },
      "end": { "line": 626, "col": 14, "offset": 32601 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 626, "col": 6, "offset": 32593 },
            "end": { "line": 626, "col": 14, "offset": 32601 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "ff37ab074556812d3f0d3ef0a375b183"
            }
          }
        }
      }
    },
    {
      "check_id": "0",
      "path": "LICENSE",
      "start": { "line": 636, "col": 21, "offset": 33083 },
      "end": { "line": 636, "col": 34, "offset": 33096 },
      "extra": {
        "metavars": {
          "X": {
            "start": { "line": 636, "col": 26, "offset": 33088 },
            "end": { "line": 636, "col": 34, "offset": 33096 },
            "abstract_content": "software",
            "unique_id": {
              "type": "AST",
              "md5sum": "ccf9c58d24cfa70cbf68103c9f1d1469"
            }
          }
        }
      }
    }
  ],
  "errors": [],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}
```

### CODE REVIEW

**Comment by mjambon** [comment]
~~The offset field is wrong (identical for start and end). Fixing this.~~ Fixed!

**Review by minusworld — APPROVED** [review]

File: spacegrep/src/lib/Semgrep.atd [hunk 1]
```diff
@@ -0,0 +1,71 @@
+(*
+   Type definitions for json communication with semgrep.
```
Comment (aryx) [inline spacegrep/src/lib/Semgrep.atd:2]: Type definitions for JSON communication with the semgrep Python wrapper.

File: spacegrep/src/lib/Semgrep.atd [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+   Inferred from JSON_report.ml in semgrep-core.
+*)
+
+type match_results = {
```
Comment (aryx) [inline spacegrep/src/lib/Semgrep.atd:7]: Nice!
I should probably use it in semgrep-core too now that they are in the same repo :)

File: spacegrep/src/lib/Semgrep.atd [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+type match_ = {
+  ?check_id: string option; (* rule ID *)
+  path: string; (* source file *)
+  start: position;
+  end_ <json name="end">: position;
+  extra: match_extra;
+}
+
+(* See R2c.ml in semgrep-core *)
```
Comment (aryx) [inline spacegrep/src/lib/Semgrep.atd:26]: it's actually in pfff/h_program-lang/R2c.ml

File: spacegrep/src/lib/Semgrep.atd [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+   In order to accommodate two record types into one variant, use
+   the atdgen feature called a json adapter.
+*)
+type unique_id = {
+  type_ <json name="type">: unique_id_type;
+  md5sum: string;
+}
+
+type unique_id_type = [
+(*  | ID <json name="id"> *) (* variant not supported at the moment *)
```
Comment (aryx) [inline spacegrep/src/lib/Semgrep.atd:69]: Anyway you should need it. It's used when a metavariable matches an identifier and we want to make sure that independent matches on the same identifer name correspond to same scoped identifier (see Naming_AST.ml in pfff))

File: spacegrep/src/lib/Semgrep.ml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    errors = [];
+    stats = {
+      okfiles = List.length doc_matches;
+      errorfiles = 0;
+    };
+  }
+
+let print_semgrep_json doc_matches =
+  make_semgrep_json doc_matches
+  |> Semgrep_j.string_of_match_results
```
Comment (aryx) [inline spacegrep/src/lib/Semgrep.ml:76]: Nice! I need to use this atdgen more.

**Review by aryx — COMMENTED** [review]
Cool!

**Pull request merged**